iotilecore/iotile/core/hw/hwmanager.py

```diff
@@ -785,7 +788,47 @@ def get_proxy(self, short_name):
         if short_name not in self._name_map:
             return None
 
-        return self._name_map[short_name][0]
+        proxy_match = self.find_correct_proxy_version(self._name_map[short_name], version)
+        print("proxy_match: ", proxy_match)
+        return proxy_match if proxy_match is not None else self._name_map[short_name][0] 
+
+    def find_correct_proxy_version(self, proxies, version):
+        """Retrieves the ModuleVersion of each proxy and match it with the tile version"""
+
+        proxy_details = {}
+        tile_version = SemanticVersion(version[0], version[1], version[2])
+        latest_version = SemanticVersion(0, 0, 0)
+
+        for proxy in proxies:
+            proxy_details[proxy] = {}
+            try:
+                # If proxy has ModuleVersion(), get the SemanticVersionRange
+                module_version = proxy.ModuleVersion()
+                proxy_details[proxy]['ModuleVersion'] = module_version
+            except AttributeError:
+                # If proxy does not have ModuleVersion(), use the name of the support package
+                stripped_proxy_version = str(proxy).split('.')[0].split('_')[-1]
+                module_version = SemanticVersionRange.FromString('^' + stripped_proxy_version + '.0.0')
+                proxy_details[proxy]['ModuleVersion'] = module_version
```

Comment:
I don't think that this will work in all cases, because the support packages do not require any versioning in their title. `FromString` might complain that it's not a parse-able version.

We probably shouldn't attempt to find a version from the support package's name. It requires us to make too many assumptions about the compatibility, and I don't think that most of our proxies would necessarily benefit from being forced to follow these assumptions. Why would we say the acceptable range is `^4.0.0` as opposed to exactly equal, or `~4.0.0`? We would have to define that somewhere, and provide hints as to why a module is not passing checks for loading proxies so that developers could debug that. 

We should be able to untangle the cases without having to hack around by modifying it as such:

- Apply `ModuleVersion()` value if present, else apply `^0.0.0` instead of trying to parse the package name

This should have the effect of our return logic iterating through the choices correctly. 

Case 1: There are multiple `ModuleVersion` entries for our proxy:
```
proxy_details = { 'proxy_one': {'ModuleVersion': SemanticVersionRange(~4.0.0)}, 'proxy_two': {'ModuleVersion': SemanticVersionRange(~3.0.0)} }

tile_version = 4.0.0
```
Then, later in this function, you iterate through the proxies and "best match" the tile_version.

Case 2: There are multiple proxies, but each of them have the same version (i.e. no version set, but someone installed two):
```
proxy_details = { 'proxy_one': {'ModuleVersion': SemanticVersionRange(0.0.0)}, 'proxy_two': {'ModuleVersion': SemanticVersionRange(0.0.0)} }
tile_version = 2.0
```
How do we do it here? The algorithm needs to properly identify that we should still pick one of the two proxies even though the tile version is larger, right? Do we need a special case to handle this (can we just return early once this is noticed)? Should we log any debug info here so someone can turn up logs when their proxy doesn't load when they expect it to?

Reply:
Agreed. Using support package name is a fickle method. As @mattrunchey and I discussed, we should try to seek a best match if `ModuleVersion` does return something useful. If it doesn't then we would pretty much have to select a proxy like we did before -- just choose one.